src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/FindAllReferences/FindAllReferencesEndpoint.cs

```diff
@@ -0,0 +1,157 @@
+            LanguageServerFeatureOptions featureOptions,
+            TagHelperFactsService tagHelperFactsService,
+            HtmlFactsService htmlFactsService)
+            : base(languageServerFeatureOptions, documentMappingService, languageServer, loggerFactory.CreateLogger<FindAllReferencesEndpoint>())
+        {
+            _featureOptions = featureOptions;
+            _documentMappingService = documentMappingService ?? throw new ArgumentNullException(nameof(documentMappingService));
+        }
+
+        public RegistrationExtensionResult GetRegistration(VSInternalClientCapabilities clientCapabilities)
+        {
+            const string AssociatedServerCapability = "referencesProvider";
+
+            var registrationOptions = new ReferenceOptions()
+            {
+                // TODO: GH issue link to support progress 
+                WorkDoneProgress = false,
+            };
+
+            return new RegistrationExtensionResult(AssociatedServerCapability, new SumType<bool, ReferenceOptions>(registrationOptions));
+        }
+
+        protected override string CustomMessageTarget => RazorLanguageServerCustomMessageTargets.RazorReferencesEndpointName;
+
+        protected override Task<IDelegatedParams?> CreateDelegatedParamsAsync(ReferenceParamsBridge request, RazorRequestContext requestContext, Projection projection, CancellationToken cancellationToken)
+        {
+            var documentContext = requestContext.GetRequiredDocumentContext();
+            return Task.FromResult<IDelegatedParams?>(new DelegatedPositionParams(
+                    documentContext.Identifier,
+                    projection.Position,
+                    projection.LanguageKind));
+        }
+
+        protected override async Task<VSInternalReferenceItem[]> HandleDelegatedResponseAsync(VSInternalReferenceItem[] delegatedResponse, ReferenceParamsBridge originalRequest, RazorRequestContext requestContext, Projection projection, CancellationToken cancellationToken)
+        {
+            var remappedLocations = new List<VSInternalReferenceItem>();
+
+            foreach (var referenceItem in delegatedResponse)
+            {
+                if (referenceItem?.Location is null || referenceItem.Text is null)
+                {
+                    continue;
+                }
+
+                // Temporary fix for codebehind leaking through
+                // Revert when https://github.com/dotnet/aspnetcore/issues/22512 is resolved
+                referenceItem.DefinitionText = FilterReferenceDisplayText(referenceItem.DefinitionText);
+                referenceItem.Text = FilterReferenceDisplayText(referenceItem.Text);
+
+                // Indicates the reference item is directly available in the code
+                referenceItem.Origin = VSInternalItemOrigin.Exact;
+
+                if (!_featureOptions.IsVirtualCSharpFile(referenceItem.Location.Uri) &&
+                    !_featureOptions.IsVirtualHtmlFile(referenceItem.Location.Uri))
```

Comment:
Does web tools support FAR? If not, I think it would be better to return null for `CreateDelegateParameters` if we're in a html context, so at least its explicitly not supported. I'm a bit worried about them adding support later, and this code starting to do weird things.

eg., I'm pretty sure the `TryMap` call below will either throw, or return false, if its asked to map from Html to Razor, but neither of those is the correct thing to do.